src/url.rs

```diff
@@ -190,6 +190,37 @@ impl PyUrl {
         }
         cls.call1((url,))
     }
+
+    #[pyo3(signature=(path, *args, no_trailing_slash=false))]
+    pub fn join(&self, path: &str, args: &Bound<'_, PyTuple>, no_trailing_slash: bool) -> PyResult<Self> {
+        let join_path = |base: &Url, partial_path: String| -> Result<Url, PyErr> {
+            base.join(&partial_path)
+                .map_err(|err| PyValueError::new_err(err.to_string()))
+        };
+
+        let add_trailing_slash = |path_str: &mut String| {
+            if !path_str.trim_end().ends_with('/') {
+                path_str.push('/');
+            }
+        };
+
+        let num_of_args = args.len();
+        let mut path_str = path.to_string();
+        if num_of_args != 0 || !no_trailing_slash {
+            add_trailing_slash(&mut path_str);
+        }
+
+        let mut new_url = join_path(&self.lib_url, path_str)?;
+
+        for i in 0..num_of_args {
+            let mut arg_str = args.get_item(i)?.to_string();
+            if i != num_of_args - 1 || !no_trailing_slash {
+                add_trailing_slash(&mut arg_str);
+            }
+            new_url = join_path(&new_url, arg_str)?;
+        }
```

Comment:
I think this might be equivalent, reads simpler to me, and only does URL parsing once:

```suggestion
        let mut url = self.lib_url.to_string();
        
        let mut add_part = |part| {
            if !url.ends_with('/') {
                url.push('/');
            }
            url.push_str(part);
        }
        
        add_part(path);
        
        for path in args {
             add_part(path)
        }
        
        if !no_trailing_slash && !url.ends_with('/') {
             url.push('/');
        }
        
        let new_url = Url::parse(url).map_err(|err| PyValueError::new_err(err.to_string()));
```

... what do you think of this implementation?

Reply:
Yes this implementation looks better, but in the existing implementation I'm joining each and every part individually to support code like `a.join("/a/b?q=a","/c")` which will result in `http://example.com/c` while the suggested implementation will result in `http://example.com/a/b?q=a/c`.
I think this would be a fair assumption that user wants to join a continuous path and will have the fragments and queries at the end of the join. So this looks good to me.
I think there should be a check for path too.
```rust
// suggested
let mut add_part = |part| {
    if !url.ends_with('/') {
        url.push('/');
    }
    url.push_str(part);
}

// change to following
let mut add_part = |part : &str| {
    if !url.ends_with('/') {
        url.push('/');
    }

    if part.starts_with('/') {
        url.push_str(&part[1..]);
    } else {
        url.push_str(part);
    }
};
```
```bash
# for path: /a/b
# suggested will return
http://example.com//a/b
# while changed will return
http://example.com/a/b
```
Shall I commit this change?

Reply:
This is a great question. In my opinion neither behaviour seems great, it seems surprising (to me) that passing an absolute url `/c` will cause all previous arguments to be disregarded. Are there precedents we can learn from (e.g. in `pathlib`, or maybe other URL handling libraries)?

Reply:
`pathlib` and `urllib.parse.urljoin` of Python and Rust's `url::Url.join` all show the same behavior where the absolute path overwrites the previous path of the URL. In my opinion, the best we can do is to provide a flag to the user `strict=false` which would join as the other libraries do, while the `strict=true` would join them as you suggested, considering the arguments strictly as part of the path.

Something similar to this,
```python
a = Url("https://example.com")
a.join("a", "/b") # https://example.com/b/
a.join("a", "/b", strict=True) # https://example.com/a/b/
```

Reply:
Hey @davidhewitt, just wanted to follow up on my last response

Reply:
Sorry for the long delay; I do recall starting a response at some point but must have never finished it.

```python
>>> Path("foo").joinpath("a", "/b")
PosixPath('/b')
```

Ok, so Pathlib is variadic in the same way _and_ as you say does overwrite the base url. So your implementation looks correct to me and accounts for the complexities.